File: backend/websocket.ts
import * as crypto from "crypto";
import { sql } from "bun";

import * as s3 from "./s3";
import { supabase } from "./supabase";
import { Packet, PacketReader } from "../util/packet";

type WebsocketData = Machine | User | undefined;

interface Machine {
  type: "machine";
  machineId: number;
}

interface User {
  type: "user";
  userId: string;
  projectId: string;
}

const onlineMachines = new Set<number>();

async function verifySignature(
  id: string,
  publicKeyPem: string,
  signature: Buffer,
): Promise<boolean> {
  const messageEncoder = new TextEncoder();
  const message = messageEncoder.encode(id);

  return crypto.verify(
    null,
    message,
    {
      key: publicKeyPem,
      format: "pem",
      type: "spki",
    },
    signature,
  );
}

async function tryMachineAuth(
  ws: Bun.ServerWebSocket<unknown>,
  message: Buffer,
): Promise<number | undefined> {
  if (message.length < 1 + 1 + 64 || message.length > 1 + 64 + 64) {
    ws.close(4001);
    return;
  }

  const idLength = message[0] as number;
  if (message.length !== 1 + idLength + 64) {
    console.log(message.length, idLength);
    ws.close(4002);
    return;
  }

  const idBuffer = message.subarray(1, 1 + idLength);
  const signature = message.subarray(1 + idLength);

  const idString = idBuffer.toString();

  const machineId = parseInt(idString, 10);
  if (isNaN(machineId)) {
    ws.close(4003);
    return;
  }

  const machineRows = await sql`
    SELECT public_key
    FROM machines
    WHERE id = ${machineId}
  `;

  if (machineRows.length === 0) {
    ws.close(1011);
    return;
  }

  const machineData = machineRows[0];
  if (typeof machineData.public_key !== "string") {
    ws.close(4004);
    return;
  }

  try {
    const isValid = await verifySignature(
      idString,
      machineData.public_key,
      signature,
    );

    if (!isValid) {
      ws.close(4005);
      return;
    }
  } catch (error) {
    console.error(error);
    ws.close(1011);
    return;
  }

  (ws.data as WebsocketData) = {
    type: "machine",
    machineId,
  };

  onlineMachines.add(machineId);
  console.log(`Machine ${machineId} authenticated from ${ws.remoteAddress}`);

  const deploymentRows = await sql`
    SELECT p.scene, d.compiled_object
    FROM machines m
    JOIN projects p ON p.id = m.project
    JOIN deployments d ON d.project_id = p.id
    WHERE m.id = ${machineId}
    ORDER BY d.created_at DESC
    LIMIT 1
  `;

  if (deploymentRows.length === 0) {
    ws.close(1011);
    return;
  }

  const project = deploymentRows[0];
  const objectId = project.compiled_object;
  if (!objectId) {
    ws.close(4006);
    return;
  }

  const scene = JSON.parse(project.scene);

  const files = [objectId, ...scene[0].promptImages];
  const urls = await Promise.all(files.map((id) => s3.presignUrl(id, 60 * 5)));
  const hashes = await Promise.all(files.map((id) => s3.getHash(id)));

  const packet = new Packet();
  packet.u8(0);
  packet.string(urls[0]!);
  packet.bytes(hashes[0]!);
  packet.u8(urls.length - 1);
  for (let i = 1; i < urls.length; i++) {
    packet.string(urls[i]!);
    packet.bytes(hashes[i]!);
  }

  ws.sendBinary(packet.toBuffer());
}

async function tryUserAuth(ws: Bun.ServerWebSocket<unknown>, message: string) {
  const parts = message.split("|");
  if (parts.length !== 2) {
    ws.close(4007);
    return;
  }

  const { data, error } = await supabase.auth.getUser(parts[0]);
  if (error) {
    console.error(error);
    ws.close(4000);
    return;
  }

  const projectRows = await sql`
    SELECT owner, scene
    FROM projects
    WHERE id = ${parts[1]!}
    LIMIT 1
  `;

  if (projectRows.length === 0) {
    ws.close(4009);
    return;
  }

  const project = projectRows[0];
  if (project.owner !== data.user.id) {
    ws.close(4008);
    return;
  }

  (ws.data as WebsocketData) = {
    type: "user",
    userId: data.user.id,
    projectId: parts[1]!,
  };

  console.log(`User ${data.user.id} authenticated from ${ws.remoteAddress}`);

  const scene = project.scene as string;
  const sceneData = JSON.parse(scene);
  ws.sendText("scene|" + scene);

  for (const object of sceneData) {
    if (object.type === "machine") {
      const machineId = object.id;
      const machineOnline = onlineMachines.has(machineId);
      ws.sendText("machineonline|" + machineId + "|" + machineOnline);
    }
  }

  for (const image of sceneData[0].promptImages) {
    const url = await s3.presignUrl(image, 60 * 5);
    const response = new Packet();
    response.u8(1);
    response.string(url);
    ws.sendBinary(response.toBuffer());
  }
}

async function handleUserMessage(
  ws: Bun.ServerWebSocket<unknown>,
  message: string | Buffer,
) {
  const userData = ws.data as User;
  if (!Buffer.isBuffer(message)) {
    ws.close(4010);
    return;
  }

  if (message.length < 1) {
    ws.close(4011);
    return;
  }

  const reader = new PacketReader(message.buffer as ArrayBuffer);
  const id = reader.u8();
  if (id === undefined) {
    ws.close(4012);
    return;
  }

  switch (id) {
    case 0:
      const fileCount = reader.u8();
      if (fileCount === undefined) {
        ws.close(4014);
        return;
      }

      for (let i = 0; i < fileCount; i++) {
        const data = reader.dynbytes();
        if (data === undefined) {
          ws.close(4015);
          return;
        }

        const projectRows = await sql`
          SELECT id, scene
          FROM projects
          WHERE id = ${userData.projectId}
        `;

        if (projectRows.length === 0) {
          ws.close(4016);
          return;
        }

        const projectData = projectRows[0];

        const id = Bun.randomUUIDv7();

        const scene = JSON.parse(projectData.scene);
        scene[0].promptImages.push(id);

        await s3.uploadFile(id, Buffer.from(data));
        const url = await s3.presignUrl(id, 60 * 5);

        await sql`
          UPDATE projects
          SET scene = ${JSON.stringify(scene)}
          WHERE id = ${userData.projectId}
        `;

        const response = new Packet();
        response.u8(1);
        response.string(url);
        ws.sendBinary(response.toBuffer());
      }
      break;

    default:
      ws.close(4013);
      return;
  }
}

export function upgradeWebsocket(
  req: Request,
  server: Bun.Server,
): Response | void {
  if (server.upgrade(req, { data: undefined })) {
    return;
  }
  return new Response("websocket upgrade failed", { status: 500 });
}

export const websocket = {
  idleTimeout: 10,
  async message(ws: Bun.ServerWebSocket<unknown>, message: string | Buffer) {
    const data = ws.data as WebsocketData;

    const notAuthenticated = data === undefined;
    if (notAuthenticated) {
      console.log(`Attempting authentication of ${ws.remoteAddress}`);
      if (typeof message === "string") {
        tryUserAuth(ws, message);
      } else {
        tryMachineAuth(ws, message);
      }
      return;
    }

    if (data.type === "machine") {
      console.log(`[machine ${data.machineId}]`, message);
    } else {
      handleUserMessage(ws, message);
    }
  },
  async close(ws: Bun.ServerWebSocket<unknown>) {
    const data = ws.data as WebsocketData;

    if (data?.type === "machine") {
      onlineMachines.delete(data.machineId);
    }
  },
};

File: backend/websocket.go
package main

import (
	"context"
	"crypto/ed25519"
	"crypto/x509"
	"database/sql"
	"encoding/json"
	"encoding/pem"
	"log"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/gorilla/websocket"
	"github.com/nedpals/supabase-go"

	"simulo.tech/backend/m/v2/protocol"
)

type WebSocketData interface {
	GetType() string
}

type MachineData struct {
	MachineID int
}

func (m *MachineData) GetType() string {
	return "machine"
}

type UserData struct {
	UserID    string
	ProjectID string
}

func (u *UserData) GetType() string {
	return "user"
}

type WebSocketHandler struct {
	server         *Server
	supabaseCli    *supabase.Client
	s3Client       *S3Client
	conn           *websocket.Conn
	onlineMachines *OnlineMachines
}

type OnlineMachines struct {
	machines map[int]bool
	mutex    sync.RWMutex
}

func NewOnlineMachines() *OnlineMachines {
	return &OnlineMachines{
		machines: make(map[int]bool),
	}
}

func (om *OnlineMachines) Add(machineID int) {
	om.mutex.Lock()
	defer om.mutex.Unlock()
	om.machines[machineID] = true
}

func (om *OnlineMachines) Remove(machineID int) {
	om.mutex.Lock()
	defer om.mutex.Unlock()
	delete(om.machines, machineID)
}

func (om *OnlineMachines) Has(machineID int) bool {
	om.mutex.RLock()
	defer om.mutex.RUnlock()
	return om.machines[machineID]
}

func NewWebSocketHandler(server *Server, supabaseCli *supabase.Client, s3Client *S3Client, conn *websocket.Conn) *WebSocketHandler {
	return &WebSocketHandler{
		server:         server,
		supabaseCli:    supabaseCli,
		s3Client:       s3Client,
		conn:           conn,
		onlineMachines: NewOnlineMachines(),
	}
}

func (ws *WebSocketHandler) Handle() {
	var data WebSocketData

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()
	go ws.pingRoutine(ctx)

	for {
		messageType, message, err := ws.conn.ReadMessage()
		if err != nil {
			log.Printf("WebSocket read error: %v", err)
			break
		}

		if data == nil {
			// Authentication phase
			if messageType == websocket.TextMessage {
				data = ws.tryUserAuth(string(message))
			} else if messageType == websocket.BinaryMessage {
				data = ws.tryMachineAuth(message)
			}

			if data == nil {
				ws.conn.Close()
				break
			}
		} else {
			// Handle authenticated messages
			switch d := data.(type) {
			case *MachineData:
				log.Printf("[machine %d] %v", d.MachineID, string(message))
			case *UserData:
				ws.handleUserMessage(d, message)
			}
		}
	}

	// Cleanup on disconnect
	if machineData, ok := data.(*MachineData); ok {
		ws.onlineMachines.Remove(machineData.MachineID)
	}
}

func (ws *WebSocketHandler) tryMachineAuth(message []byte) WebSocketData {
	if len(message) < 1+1+64 || len(message) > 1+64+64 {
		ws.conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(4001, "invalid message length"))
		return nil
	}

	idLength := int(message[0])
	if len(message) != 1+idLength+64 {
		ws.conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(4002, "message length mismatch"))
		return nil
	}

	idBuffer := message[1 : 1+idLength]
	signature := message[1+idLength:]
	idString := string(idBuffer)

	machineID, err := strconv.Atoi(idString)
	if err != nil {
		ws.conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(4003, "invalid machine ID"))
		return nil
	}

	query := "SELECT public_key FROM machines WHERE id = $1"

	var publicKey string
	err = ws.server.db.QueryRow(query, machineID).Scan(&publicKey)
	if err != nil {
		if err == sql.ErrNoRows {
			ws.conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(4005, "not authorized"))
			return nil
		}
		ws.conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(1011, "machine not found"))
		return nil
	}

	if !ws.verifySignature(idString, publicKey, signature) {
		ws.conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(4005, "not authorized"))
		return nil
	}

	ws.onlineMachines.Add(machineID)
	log.Printf("Machine %d authenticated", machineID)

	ws.sendMachineProject(machineID)

	return &MachineData{MachineID: machineID}
}

func (ws *WebSocketHandler) tryUserAuth(message string) WebSocketData {
	parts := strings.Split(message, "|")
	if len(parts) != 2 {
		ws.conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(4007, "invalid auth format"))
		return nil
	}

	user, err := ws.supabaseCli.Auth.User(context.Background(), parts[0])
	if err != nil {
		ws.conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(4000, "unauthorized"))
		return nil
	}

	projectData, err := ws.server.GetProject(parts[1])
	if err != nil {
		ws.conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(4009, "project not found"))
		return nil
	}

	if projectData.Owner != user.ID {
		ws.conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(4008, "not project owner"))
		return nil
	}

	log.Printf("User %s authenticated", user.ID)

	// Send scene data
	ws.conn.WriteMessage(websocket.TextMessage, protocol.S2EInitScene(projectData.Scene))

	// Send machine online status
	var sceneData []map[string]interface{}
	json.Unmarshal([]byte(projectData.Scene), &sceneData)

	for _, object := range sceneData {
		if object["type"] == "machine" {
			if machineID, ok := object["id"].(float64); ok {
				online := ws.onlineMachines.Has(int(machineID))
				ws.conn.WriteMessage(websocket.TextMessage, protocol.S2EMachineOnline(int(machineID), online))
			}
		}
	}

	if len(sceneData) > 0 {
		if promptImages, ok := sceneData[0]["promptImages"].([]interface{}); ok {
			for _, imageID := range promptImages {
				if idStr, ok := imageID.(string); ok {
					url, err := ws.s3Client.PresignURL(idStr, 5*time.Minute)
					if err == nil {
						ws.conn.WriteMessage(websocket.BinaryMessage, protocol.S2EAddPromptImage(url))
					}
				}
			}
		}
	}

	return &UserData{UserID: user.ID, ProjectID: parts[1]}
}

func (ws *WebSocketHandler) handleUserMessage(userData *UserData, message []byte) {
	if len(message) < 1 {
		ws.conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(4011, "empty message"))
		return
	}

	reader := protocol.NewPacketReader(message)
	id, err := reader.U8()
	if err != nil {
		ws.conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(4012, "invalid packet"))
		return
	}

	switch id {
	case protocol.E2SAddImagesId:
		var packet protocol.E2SAddImages
		if err := packet.Unmarshal(reader); err != nil {
			ws.conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(4014, "protocol error"))
			return
		}

		for _, data := range packet.Uploads {
			projectData, err := ws.server.GetProject(userData.ProjectID)
			if err != nil {
				ws.conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(4016, "project not found"))
				return
			}

			// Generate new UUID for the file
			fileID := generateRandomHex(16)

			// Parse and update scene
			var scene []map[string]interface{}
			json.Unmarshal([]byte(projectData.Scene), &scene)

			if len(scene) > 0 {
				if promptImages, ok := scene[0]["promptImages"].([]interface{}); ok {
					scene[0]["promptImages"] = append(promptImages, fileID)
				} else {
					scene[0]["promptImages"] = []interface{}{fileID}
				}
			}

			// Upload to S3
			if err := ws.s3Client.UploadBuffer(fileID, data); err != nil {
				log.Printf("S3 upload failed: %v", err)
				continue
			}

			// Update database
			updatedScene, _ := json.Marshal(scene)
			ws.server.UpdateProjectScene(userData.ProjectID, string(updatedScene))

			// Send presigned URL back
			url, err := ws.s3Client.PresignURL(fileID, 5*time.Minute)
			if err == nil {
				packet := protocol.NewPacket()
				packet.U8(1)
				packet.String(url)
				ws.conn.WriteMessage(websocket.BinaryMessage, packet.ToBuffer())
			}
		}

	case protocol.E2SDeleteImageId:
		var packet protocol.E2SDeleteImage
		if err := packet.Unmarshal(reader); err != nil {
			ws.conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(4014, "protocol error"))
			return
		}

		projectData, err := ws.server.GetProject(userData.ProjectID)
		if err != nil {
			ws.conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(4016, "project not found"))
			return
		}

		scene := []map[string]interface{}{}
		json.Unmarshal([]byte(projectData.Scene), &scene)

		if len(scene) > 0 {
			promptImages, ok := scene[0]["promptImages"].([]interface{})
			if ok {
				index := int(packet.Index)
				if index < len(promptImages) {
					_ = ws.s3Client.Delete(promptImages[index].(string))
					scene[0]["promptImages"] = append(promptImages[:index], promptImages[index+1:]...)
				}
			}
		}

		updatedScene, _ := json.Marshal(scene)
		ws.server.UpdateProjectScene(userData.ProjectID, string(updatedScene))

		ws.conn.WriteMessage(websocket.BinaryMessage, protocol.S2EDeletePromptImage(packet.Index))

	default:
		ws.conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(4013, "unknown message type"))
	}
}

func (ws *WebSocketHandler) sendMachineProject(machineID int) {
	query := `
		SELECT object, name
		FROM project_assets
		WHERE project = (
			SELECT project
			FROM machines
			WHERE id = $1
		)
	`

	rows, err := ws.server.db.Query(query, machineID)
	if err != nil {
		log.Printf("Failed to get project data: %v", err)
		return
	}
	defer rows.Close()

	var programUrl string
	var programHash []byte
	urls := []string{}
	names := []string{}
	hashes := [][]byte{}

	for rows.Next() {
		var object, name string
		if err := rows.Scan(&object, &name); err != nil {
			log.Printf("Failed to scan row: %v", err)
			return
		}

		url, err := ws.s3Client.PresignURL(object, 5*time.Minute)
		if err != nil {
			log.Printf("Failed to presign URL for %s: %v", object, err)
			return
		}

		s3Hash, err := ws.s3Client.GetHash(object)
		if err != nil {
			log.Printf("Failed to get hash for %s: %v", object, err)
			return
		}

		if name == "main.wasm" {
			programUrl = url
			programHash = s3Hash
		} else {
			urls = append(urls, url)
			names = append(names, name)
			hashes = append(hashes, s3Hash)
		}
	}

	ws.conn.WriteMessage(websocket.BinaryMessage, protocol.S2MInitAssets(programUrl, programHash, names, urls, hashes))
}

func (ws *WebSocketHandler) pingRoutine(ctx context.Context) {
	ticker := time.NewTicker(30 * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			if err := ws.conn.WriteMessage(websocket.PingMessage, nil); err != nil {
				log.Printf("WebSocket ping error: %v", err)
				return
			}
		}
	}
}

func (ws *WebSocketHandler) verifySignature(id, publicKeyPem string, signature []byte) bool {
	block, _ := pem.Decode([]byte(publicKeyPem))
	if block == nil {
		return false
	}

	publicKey, err := x509.ParsePKIXPublicKey(block.Bytes)
	if err != nil {
		return false
	}

	ed25519PubKey, ok := publicKey.(ed25519.PublicKey)
	if !ok {
		return false
	}

	message := []byte(id)
	return ed25519.Verify(ed25519PubKey, message, signature)
}
